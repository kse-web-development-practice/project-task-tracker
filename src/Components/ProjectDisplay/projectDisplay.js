import React, { useState } from 'react'
import PropTypes from 'prop-types'
import styles from './projectDisplay.module.css'
import { ListCell } from '../ListCell/listCell'
import { ListSelect } from '../ListSelect/listSelect'

const statusOptions = ['Completed', 'In progress', 'Not started']

export const ProjectDisplay = ({ name, totalTasks, completedTasks, initialStatus }) => {
  const [status, setStatus] = useState(initialStatus)

  const handleStatusChange = (e) => {
    setStatus(e.target.value)
  }

  return (
    <tr className={styles.projectDisplay}>
      <ListCell>{name}</ListCell>
      <ListCell>
        {completedTasks}/{totalTasks}
      </ListCell>
      <ListCell>
        <ListSelect value={status} options={statusOptions} onChange={handleStatusChange} />
      </ListCell>
    </tr>
  )
}

ProjectDisplay.propTypes = {
  name: PropTypes.string.isRequired,
  initialStatus: PropTypes.oneOf(statusOptions).isRequired,
  totalTasks: PropTypes.number.isRequired,
  completedTasks: PropTypes.number.isRequired
}
