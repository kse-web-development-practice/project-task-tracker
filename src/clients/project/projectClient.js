async function addProject(name, description, status, ownerUsername) {
  const project = {
    name,
    description,
    status,
    ownerUsername
  }

  const response = await fetch(`${process.env.API_BASE_URL}/project/add`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(project)
  })

  if (response.status == 500) {
    throw new Error(`Error: ${response.statusText}`)
  }

  if (!response.ok) {
    return null
  }

  const body = await response.json()
  return body.project
}

async function getProjectsByUser(ownerUsername) {
  const response = await fetch(`${process.env.API_BASE_URL}/project/${ownerUsername}`, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json'
    }
  })

  if (response.status == 500) {
    throw new Error(`Error: ${response.statusText}`)
  }

  if (!response.ok) {
    return null
  }

  const body = await response.json()
  return body.projects
}

export default {
  addProject,
  getProjectsByUser
}
