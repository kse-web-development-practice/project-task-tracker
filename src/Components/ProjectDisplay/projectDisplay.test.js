import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import { ProjectDisplay } from './projectDisplay'
import projectClient from '../../clients/project/projectClient'
import { useNavigate } from 'react-router-dom'

jest.mock('../../clients/project/projectClient', () => ({
  updateProject: jest.fn()
}))

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn()
}))

describe('ProjectDisplay Component', () => {
  const renderProjectDisplay = () =>
    render(
      <table>
        <tbody>
          <ProjectDisplay
            id="1"
            name="Project 1"
            totalTasks={10}
            completedTasks={5}
            initialStatus="In progress"
          />
        </tbody>
      </table>
    )

  test('snapshot', () => {
    const { asFragment } = renderProjectDisplay()
    expect(asFragment()).toMatchSnapshot()
  })

  test('status change', () => {
    renderProjectDisplay()

    fireEvent.change(screen.getByRole('combobox'), { target: { value: 'Completed' } })
    expect(projectClient.updateProject).toHaveBeenCalledWith('1', { status: 'Completed' })
  })

  test('click on project', () => {
    const mockNavigate = jest.fn()
    useNavigate.mockReturnValue(mockNavigate)

    renderProjectDisplay()

    fireEvent.click(screen.getByText('Project 1'))
    expect(mockNavigate).toHaveBeenCalledWith('/tasks/1')
  })
})
