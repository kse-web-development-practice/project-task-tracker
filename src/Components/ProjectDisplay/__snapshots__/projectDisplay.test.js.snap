// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectDisplay Component snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr
        class="projectDisplay"
      >
        <td
          class="listCell clickable"
        >
          Project 1
        </td>
        <td
          class="listCell"
        >
          5/10
        </td>
        <td
          class="listCell"
        >
          <select
            class="listSelect"
          >
            <option
              value="Completed"
            >
              Completed
            </option>
            <option
              value="In progress"
            >
              In progress
            </option>
            <option
              value="Not started"
            >
              Not started
            </option>
          </select>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
